In [1]:
import numpy as np
import pandas as pd

In [2]:
train_set = pd.read_csv('~/SFCrime/data/train.csv', parse_dates=['Dates'])

In [3]:
train_set.head()

Dates        Category                      Descript  \
0 2015-05-13 23:53:00        WARRANTS                WARRANT ARREST   
1 2015-05-13 23:53:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
2 2015-05-13 23:33:00  OTHER OFFENSES      TRAFFIC VIOLATION ARREST   
3 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   
4 2015-05-13 23:30:00   LARCENY/THEFT  GRAND THEFT FROM LOCKED AUTO   

   DayOfWeek PdDistrict      Resolution                    Address  \
0  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
1  Wednesday   NORTHERN  ARREST, BOOKED         OAK ST / LAGUNA ST   
2  Wednesday   NORTHERN  ARREST, BOOKED  VANNESS AV / GREENWICH ST   
3  Wednesday   NORTHERN            NONE   1500 Block of LOMBARD ST   
4  Wednesday       PARK            NONE  100 Block of BRODERICK ST   

            X          Y  
0 -122.425892  37.774599  
1 -122.425892  37.774599  
2 -122.424363  37.800414  
3 -122.426995  37.800873  
4 -122.438738  37.771541

## Explore each coloumn of train set

### Dates

In [4]:
train_set.Dates.describe()

count                  878049
unique                 389257
top       2011-01-01 00:01:00
freq                      185
first     2003-01-06 00:01:00
last      2015-05-13 23:53:00
Name: Dates, dtype: object

### Category

In [5]:
train_set.Category.describe()

count            878049
unique               39
top       LARCENY/THEFT
freq             174900
Name: Category, dtype: object

In [6]:
train_set.Category.value_counts()

LARCENY/THEFT                  174900
OTHER OFFENSES                 126182
NON-CRIMINAL                    92304
ASSAULT                         76876
DRUG/NARCOTIC                   53971
VEHICLE THEFT                   53781
VANDALISM                       44725
WARRANTS                        42214
BURGLARY                        36755
SUSPICIOUS OCC                  31414
MISSING PERSON                  25989
ROBBERY                         23000
FRAUD                           16679
FORGERY/COUNTERFEITING          10609
SECONDARY CODES                  9985
WEAPON LAWS                      8555
PROSTITUTION                     7484
TRESPASS                         7326
STOLEN PROPERTY                  4540
SEX OFFENSES FORCIBLE            4388
DISORDERLY CONDUCT               4320
DRUNKENNESS                      4280
RECOVERED VEHICLE                3138
KIDNAPPING                       2341
DRIVING UNDER THE INFLUENCE      2268
RUNAWAY                          1946
LIQUOR LAWS 

### Descript

In [7]:
train_set.Descript.describe()

count                           878049
unique                             879
top       GRAND THEFT FROM LOCKED AUTO
freq                             60022
Name: Descript, dtype: object

### DayOfWeek

In [8]:
train_set.DayOfWeek.value_counts()

Friday       133734
Wednesday    129211
Saturday     126810
Thursday     125038
Tuesday      124965
Monday       121584
Sunday       116707
Name: DayOfWeek, dtype: int64

### PdDistrict

In [9]:
train_set.PdDistrict.value_counts()

SOUTHERN      157182
MISSION       119908
NORTHERN      105296
BAYVIEW        89431
CENTRAL        85460
TENDERLOIN     81809
INGLESIDE      78845
TARAVAL        65596
PARK           49313
RICHMOND       45209
Name: PdDistrict, dtype: int64

### Resolution

In [10]:
train_set.Resolution.describe()

count     878049
unique        17
top         NONE
freq      526790
Name: Resolution, dtype: object

In [11]:
train_set.Resolution.value_counts()

NONE                                      526790
ARREST, BOOKED                            206403
ARREST, CITED                              77004
LOCATED                                    17101
PSYCHOPATHIC CASE                          14534
UNFOUNDED                                   9585
JUVENILE BOOKED                             5564
COMPLAINANT REFUSES TO PROSECUTE            3976
DISTRICT ATTORNEY REFUSES TO PROSECUTE      3934
NOT PROSECUTED                              3714
JUVENILE CITED                              3332
PROSECUTED BY OUTSIDE AGENCY                2504
EXCEPTIONAL CLEARANCE                       1530
JUVENILE ADMONISHED                         1455
JUVENILE DIVERTED                            355
CLEARED-CONTACT JUVENILE FOR MORE INFO       217
PROSECUTED FOR LESSER OFFENSE                 51
Name: Resolution, dtype: int64

### Address

In [12]:
train_set.Address.describe()

count                     878049
unique                     23228
top       800 Block of BRYANT ST
freq                       26533
Name: Address, dtype: object

### X and Y

In [13]:
train_set.X.describe()

count    878049.000000
mean       -122.422616
std           0.030354
min        -122.513642
25%        -122.432952
50%        -122.416420
75%        -122.406959
max        -120.500000
Name: X, dtype: float64

In [14]:
train_set.Y.describe()

count    878049.000000
mean         37.771020
std           0.456893
min          37.707879
25%          37.752427
50%          37.775421
75%          37.784369
max          90.000000
Name: Y, dtype: float64

#### There is some bad data in Y.

In [15]:
train_set[train_set.Y > 40]

Dates               Category  \
660485 2005-12-30 17:00:00          LARCENY/THEFT   
660711 2005-12-30 00:34:00                ASSAULT   
660712 2005-12-30 00:34:00                ASSAULT   
661106 2005-12-29 00:07:00           NON-CRIMINAL   
666430 2005-11-30 11:25:00         OTHER OFFENSES   
667042 2005-11-28 16:04:00               TRESPASS   
669946 2005-11-14 09:20:00         OTHER OFFENSES   
671709 2005-11-02 20:30:00         OTHER OFFENSES   
673078 2005-10-23 20:00:00          LARCENY/THEFT   
673114 2005-10-23 18:11:00               WARRANTS   
673115 2005-10-23 18:11:00         OTHER OFFENSES   
673510 2005-10-22 16:00:00          LARCENY/THEFT   
674043 2005-10-21 10:15:00          LARCENY/THEFT   
675070 2005-10-18 20:00:00          VEHICLE THEFT   
675073 2005-10-18 20:00:00          VEHICLE THEFT   
675288 2005-10-18 10:15:00         SUSPICIOUS OCC   
675540 2005-10-17 18:20:00              VANDALISM   
677036 2005-10-07 08:35:00           NON-CRIMINAL   
677260 2005-10-06 17:30:00          VEHICLE THEFT   
677268 2005-10-06 17:30:00          VEHICLE THEFT   
677828 2005-10-05 12:00:00          LARCENY/THEFT   
678326 2005-10-04 09:00:00          LARCENY/THEFT   
678539 2005-10-03 18:46:00         OTHER OFFENSES   
678540 2005-10-03 18:46:00           NON-CRIMINAL   
679643 2005-09-23 23:00:00               BURGLARY   
682780 2005-09-09 00:03:00                ASSAULT   
683862 2005-09-06 08:12:00         OTHER OFFENSES   
684406 2005-08-28 16:00:00           NON-CRIMINAL   
685719 2005-08-25 09:26:00               WARRANTS   
685720 2005-08-25 09:26:00               WARRANTS   
...                    ...                    ...   
727080 2005-01-25 21:32:00         OTHER OFFENSES   
754828 2004-08-21 23:35:00          LARCENY/THEFT   
756059 2004-08-19 03:43:00         OTHER OFFENSES   
759292 2004-08-04 01:12:00               WARRANTS   
759397 2004-08-03 19:30:00          LARCENY/THEFT   
767318 2004-06-24 13:00:00         OTHER OFFENSES   
767501 2004-06-24 00:03:00     DISORDERLY CONDUCT   
767502 2004-06-24 00:03:00         OTHER OFFENSES   
772415 2004-05-28 18:15:00         SUSPICIOUS OCC   
772431 2004-05-28 18:00:00          VEHICLE THEFT   
773159 2004-05-27 00:45:00               WARRANTS   
783919 2004-04-03 20:52:00         OTHER OFFENSES   
788574 2004-03-17 01:16:00     DISORDERLY CONDUCT   
788575 2004-03-17 01:16:00         OTHER OFFENSES   
789255 2004-03-15 15:30:00                ASSAULT   
798834 2004-01-24 14:00:00          LARCENY/THEFT   
823378 2003-09-21 13:00:00          LARCENY/THEFT   
833362 2003-08-07 15:45:00               WARRANTS   
833921 2003-08-06 12:00:00          LARCENY/THEFT   
839585 2003-07-08 19:22:00               WARRANTS   
842259 2003-06-25 07:30:00          VEHICLE THEFT   
842261 2003-06-25 07:30:00          VEHICLE THEFT   
842434 2003-06-24 18:35:00        STOLEN PROPERTY   
842435 2003-06-24 18:35:00               WARRANTS   
842828 2003-06-23 19:00:00          LARCENY/THEFT   
844995 2003-06-11 08:49:00         OTHER OFFENSES   
845842 2003-06-09 09:25:00         OTHER OFFENSES   
852880 2003-05-02 01:00:00  SEX OFFENSES FORCIBLE   
857248 2003-04-14 16:30:00                ROBBERY   
871198 2003-02-05 12:00:00          LARCENY/THEFT   

                                         Descript  DayOfWeek  PdDistrict  \
660485               GRAND THEFT FROM LOCKED AUTO     Friday  TENDERLOIN   
660711                INFLICT INJURY ON COHABITEE     Friday     BAYVIEW   
660712       AGGRAVATED ASSAULT WITH BODILY FORCE     Friday     BAYVIEW   
661106               AIDED CASE, MENTAL DISTURBED   Thursday  TENDERLOIN   
666430                          TRAFFIC VIOLATION  Wednesday  TENDERLOIN   
667042                                TRESPASSING     Monday  TENDERLOIN   
669946      DRIVERS LICENSE, SUSPENDED OR REVOKED     Monday     BAYVIEW   
671709                MISCELLANEOUS INVESTIGATION  Wednesday     CENTRAL   
673078               GRAND THEFT FROM LOCKED AUTO 

In [16]:
train_set[train_set.Y < 40].Y.describe()

count    877982.000000
mean         37.767035
std           0.024165
min          37.707879
25%          37.752427
50%          37.775421
75%          37.784368
max          37.819975
Name: Y, dtype: float64

In [31]:
train_set[train_set.X > -121]

Dates               Category  \
660485 2005-12-30 17:00:00          LARCENY/THEFT   
660711 2005-12-30 00:34:00                ASSAULT   
660712 2005-12-30 00:34:00                ASSAULT   
661106 2005-12-29 00:07:00           NON-CRIMINAL   
666430 2005-11-30 11:25:00         OTHER OFFENSES   
667042 2005-11-28 16:04:00               TRESPASS   
669946 2005-11-14 09:20:00         OTHER OFFENSES   
671709 2005-11-02 20:30:00         OTHER OFFENSES   
673078 2005-10-23 20:00:00          LARCENY/THEFT   
673114 2005-10-23 18:11:00               WARRANTS   
673115 2005-10-23 18:11:00         OTHER OFFENSES   
673510 2005-10-22 16:00:00          LARCENY/THEFT   
674043 2005-10-21 10:15:00          LARCENY/THEFT   
675070 2005-10-18 20:00:00          VEHICLE THEFT   
675073 2005-10-18 20:00:00          VEHICLE THEFT   
675288 2005-10-18 10:15:00         SUSPICIOUS OCC   
675540 2005-10-17 18:20:00              VANDALISM   
677036 2005-10-07 08:35:00           NON-CRIMINAL   
677260 2005-10-06 17:30:00          VEHICLE THEFT   
677268 2005-10-06 17:30:00          VEHICLE THEFT   
677828 2005-10-05 12:00:00          LARCENY/THEFT   
678326 2005-10-04 09:00:00          LARCENY/THEFT   
678539 2005-10-03 18:46:00         OTHER OFFENSES   
678540 2005-10-03 18:46:00           NON-CRIMINAL   
679643 2005-09-23 23:00:00               BURGLARY   
682780 2005-09-09 00:03:00                ASSAULT   
683862 2005-09-06 08:12:00         OTHER OFFENSES   
684406 2005-08-28 16:00:00           NON-CRIMINAL   
685719 2005-08-25 09:26:00               WARRANTS   
685720 2005-08-25 09:26:00               WARRANTS   
...                    ...                    ...   
727080 2005-01-25 21:32:00         OTHER OFFENSES   
754828 2004-08-21 23:35:00          LARCENY/THEFT   
756059 2004-08-19 03:43:00         OTHER OFFENSES   
759292 2004-08-04 01:12:00               WARRANTS   
759397 2004-08-03 19:30:00          LARCENY/THEFT   
767318 2004-06-24 13:00:00         OTHER OFFENSES   
767501 2004-06-24 00:03:00     DISORDERLY CONDUCT   
767502 2004-06-24 00:03:00         OTHER OFFENSES   
772415 2004-05-28 18:15:00         SUSPICIOUS OCC   
772431 2004-05-28 18:00:00          VEHICLE THEFT   
773159 2004-05-27 00:45:00               WARRANTS   
783919 2004-04-03 20:52:00         OTHER OFFENSES   
788574 2004-03-17 01:16:00     DISORDERLY CONDUCT   
788575 2004-03-17 01:16:00         OTHER OFFENSES   
789255 2004-03-15 15:30:00                ASSAULT   
798834 2004-01-24 14:00:00          LARCENY/THEFT   
823378 2003-09-21 13:00:00          LARCENY/THEFT   
833362 2003-08-07 15:45:00               WARRANTS   
833921 2003-08-06 12:00:00          LARCENY/THEFT   
839585 2003-07-08 19:22:00               WARRANTS   
842259 2003-06-25 07:30:00          VEHICLE THEFT   
842261 2003-06-25 07:30:00          VEHICLE THEFT   
842434 2003-06-24 18:35:00        STOLEN PROPERTY   
842435 2003-06-24 18:35:00               WARRANTS   
842828 2003-06-23 19:00:00          LARCENY/THEFT   
844995 2003-06-11 08:49:00         OTHER OFFENSES   
845842 2003-06-09 09:25:00         OTHER OFFENSES   
852880 2003-05-02 01:00:00  SEX OFFENSES FORCIBLE   
857248 2003-04-14 16:30:00                ROBBERY   
871198 2003-02-05 12:00:00          LARCENY/THEFT   

                                         Descript  DayOfWeek  PdDistrict  \
660485               GRAND THEFT FROM LOCKED AUTO     Friday  TENDERLOIN   
660711                INFLICT INJURY ON COHABITEE     Friday     BAYVIEW   
660712       AGGRAVATED ASSAULT WITH BODILY FORCE     Friday     BAYVIEW   
661106               AIDED CASE, MENTAL DISTURBED   Thursday  TENDERLOIN   
666430                          TRAFFIC VIOLATION  Wednesday  TENDERLOIN   
667042                                TRESPASSING     Monday  TENDERLOIN   
669946      DRIVERS LICENSE, SUSPENDED OR REVOKED     Monday     BAYVIEW   
671709                MISCELLANEOUS INVESTIGATION  Wednesday     CENTRAL   
673078               GRAND THEFT FROM LOCKED AUTO 

In [33]:
train_set[train_set.X < -121].X.describe()

count    877982.000000
mean       -122.422763
std           0.025285
min        -122.513642
25%        -122.432952
50%        -122.416420
75%        -122.406959
max        -122.364937
Name: X, dtype: float64

#### Will finish up the exploring of test set before deciding on what to do with special X and Y values in the test set

## Explore test set

In [17]:
test_set = pd.read_csv('~/SFCrime/data/test.csv', parse_dates=['Dates'])

In [18]:
test_set.head()

Id               Dates DayOfWeek PdDistrict                   Address  \
0   0 2015-05-10 23:59:00    Sunday    BAYVIEW   2000 Block of THOMAS AV   
1   1 2015-05-10 23:51:00    Sunday    BAYVIEW        3RD ST / REVERE AV   
2   2 2015-05-10 23:50:00    Sunday   NORTHERN    2000 Block of GOUGH ST   
3   3 2015-05-10 23:45:00    Sunday  INGLESIDE  4700 Block of MISSION ST   
4   4 2015-05-10 23:45:00    Sunday  INGLESIDE  4700 Block of MISSION ST   

            X          Y  
0 -122.399588  37.735051  
1 -122.391523  37.732432  
2 -122.426002  37.792212  
3 -122.437394  37.721412  
4 -122.437394  37.721412

### Id

In [19]:
test_set.Id.describe()

count    884262.000000
mean     442130.500000
std      255264.596206
min           0.000000
25%      221065.250000
50%      442130.500000
75%      663195.750000
max      884261.000000
Name: Id, dtype: float64

### Dates

In [20]:
test_set.Dates.describe()

count                  884262
unique                 392173
top       2010-01-01 00:01:00
freq                      150
first     2003-01-01 00:01:00
last      2015-05-10 23:59:00
Name: Dates, dtype: object

### DayOfWeek

In [21]:
test_set.DayOfWeek.value_counts()

Friday       134703
Wednesday    130399
Saturday     127038
Tuesday      126940
Thursday     126541
Monday       122226
Sunday       116415
Name: DayOfWeek, dtype: int64

### PdDistrict

In [22]:
test_set.PdDistrict.value_counts()

SOUTHERN      157456
MISSION       120449
NORTHERN      107017
BAYVIEW        89591
CENTRAL        86130
TENDERLOIN     81747
INGLESIDE      80084
TARAVAL        66617
PARK           50199
RICHMOND       44972
Name: PdDistrict, dtype: int64

### Address

In [23]:
test_set.Address.describe()

count                     884262
unique                     23184
top       800 Block of BRYANT ST
freq                       26984
Name: Address, dtype: object

### X and Y

In [24]:
test_set.X.describe()

count    884262.000000
mean       -122.422693
std           0.030985
min        -122.513642
25%        -122.433069
50%        -122.416517
75%        -122.406959
max        -120.500000
Name: X, dtype: float64

In [25]:
test_set.Y.describe()

count    884262.000000
mean         37.771476
std           0.484824
min          37.707879
25%          37.752374
50%          37.775421
75%          37.784353
max          90.000000
Name: Y, dtype: float64

In [26]:
test_set[test_set.Y > 40]

Id               Dates  DayOfWeek  PdDistrict  \
664733  664733 2005-12-20 14:57:00    Tuesday  TENDERLOIN   
664734  664734 2005-12-20 14:57:00    Tuesday  TENDERLOIN   
664826  664826 2005-12-20 07:36:00    Tuesday    RICHMOND   
664895  664895 2005-12-19 23:08:00     Monday  TENDERLOIN   
667226  667226 2005-12-06 09:40:00    Tuesday    RICHMOND   
671467  671467 2005-11-10 16:30:00   Thursday  TENDERLOIN   
672453  672453 2005-11-08 06:30:00    Tuesday     CENTRAL   
675611  675611 2005-10-24 00:03:00     Monday  TENDERLOIN   
675858  675858 2005-10-16 10:27:00     Sunday  TENDERLOIN   
676528  676528 2005-10-14 15:49:00     Friday  TENDERLOIN   
676811  676811 2005-10-13 22:35:00   Thursday    SOUTHERN   
677355  677355 2005-10-12 16:00:00  Wednesday        PARK   
678944  678944 2005-10-01 16:18:00   Saturday  TENDERLOIN   
678945  678945 2005-10-01 16:18:00   Saturday  TENDERLOIN   
679267  679267 2005-09-30 23:01:00     Friday    NORTHERN   
683721  683721 2005-09-12 17:30:00     Monday  TENDERLOIN   
683795  683795 2005-09-12 13:53:00     Monday  TENDERLOIN   
683968  683968 2005-09-04 23:30:00     Sunday    NORTHERN   
683970  683970 2005-09-04 23:30:00     Sunday    NORTHERN   
686013  686013 2005-08-30 20:15:00    Tuesday  TENDERLOIN   
686014  686014 2005-08-30 20:15:00    Tuesday  TENDERLOIN   
686015  686015 2005-08-30 20:15:00    Tuesday  TENDERLOIN   
687015  687015 2005-08-21 05:30:00     Sunday    NORTHERN   
687357  687357 2005-08-20 09:30:00   Saturday  TENDERLOIN   
688886  688886 2005-08-16 09:25:00    Tuesday  TENDERLOIN   
689209  689209 2005-08-15 12:00:00     Monday  TENDERLOIN   
693007  693007 2005-07-22 22:15:00     Friday     TARAVAL   
705820  705820 2005-05-23 02:00:00     Monday     CENTRAL   
707005  707005 2005-05-13 00:18:00     Friday     CENTRAL   
709457  709457 2005-04-29 18:55:00     Friday  TENDERLOIN   
...        ...                 ...        ...         ...   
790277  790277 2004-03-27 12:48:00   Saturday    NORTHERN   
790278  790278 2004-03-27 12:48:00   Saturday    NORTHERN   
796525  796525 2004-02-27 16:39:00     Friday    NORTHERN   
796526  796526 2004-02-27 16:39:00     Friday    NORTHERN   
798831  798831 2004-02-15 02:00:00     Sunday    SOUTHERN   
798832  798832 2004-02-15 02:00:00     Sunday    SOUTHERN   
798941  798941 2004-02-14 21:10:00   Saturday    RICHMOND   
798942  798942 2004-02-14 21:10:00   Saturday    RICHMOND   
798943  798943 2004-02-14 21:10:00   Saturday    RICHMOND   
800668  800668 2004-02-10 20:30:00    Tuesday    RICHMOND   
800670  800670 2004-02-10 20:30:00    Tuesday    RICHMOND   
807432  807432 2004-01-12 05:30:00     Monday    SOUTHERN   
807434  807434 2004-01-12 05:30:00     Monday    SOUTHERN   
814294  814294 2003-12-03 16:30:00  Wednesday    SOUTHERN   
822978  822978 2003-10-22 01:38:00  Wednesday    NORTHERN   
822979  822979 2003-10-22 01:38:00  Wednesday    NORTHERN   
822980  822980 2003-10-22 01:38:00  Wednesday    NORTHERN   
822981  822981 2003-10-22 01:38:00  Wednesday    NORTHERN   
824015  824015 2003-10-12 17:46:00     Sunday    NORTHERN   
824016  824016 2003-10-12 17:46:00     Sunday    NORTHERN   
824017  824017 2003-10-12 17:46:00     Sunday    NORTHERN   
825266  825266 2003-10-09 14:51:00   Thursday     MISSION   
825915  825915 2003-10-07 23:45:00    Tuesday     BAYVIEW   
827026  827026 2003-09-28 10:00:00     Sunday    SOUTHERN   
829626  829626 2003-09-22 13:50:00     Monday   INGLESIDE   
830647  830647 2003-09-12 23:00:00     Friday    NORTHERN   
830648  830648 2003-09-12 23:00:00     Friday    NORTHERN   
836254  836254 2003-08-16 17:30:00   Saturday    SOUTHERN   
836255  836255 2003-08-16 17:30:00   Saturday    SOUTHERN   
861864  861864 2003-04-12 11:30:00   Saturday     TARAVAL   

                                          Address      X   Y  
664733              7THSTNORTH ST / MCALLISTER ST -120.5  90  
664734              7THSTNORTH ST / MCALLISTER ST -120.5  90  
664826             GEARY BL / AVENUE OF 

In [27]:
test_set[test_set.Y < 40].Y.describe()

count    884186.000000
mean         37.766986
std           0.024179
min          37.707879
25%          37.752363
50%          37.775421
75%          37.784353
max          37.820621
Name: Y, dtype: float64

In [35]:
test_set[test_set.X > -121]

Id               Dates  DayOfWeek  PdDistrict  \
664733  664733 2005-12-20 14:57:00    Tuesday  TENDERLOIN   
664734  664734 2005-12-20 14:57:00    Tuesday  TENDERLOIN   
664826  664826 2005-12-20 07:36:00    Tuesday    RICHMOND   
664895  664895 2005-12-19 23:08:00     Monday  TENDERLOIN   
667226  667226 2005-12-06 09:40:00    Tuesday    RICHMOND   
671467  671467 2005-11-10 16:30:00   Thursday  TENDERLOIN   
672453  672453 2005-11-08 06:30:00    Tuesday     CENTRAL   
675611  675611 2005-10-24 00:03:00     Monday  TENDERLOIN   
675858  675858 2005-10-16 10:27:00     Sunday  TENDERLOIN   
676528  676528 2005-10-14 15:49:00     Friday  TENDERLOIN   
676811  676811 2005-10-13 22:35:00   Thursday    SOUTHERN   
677355  677355 2005-10-12 16:00:00  Wednesday        PARK   
678944  678944 2005-10-01 16:18:00   Saturday  TENDERLOIN   
678945  678945 2005-10-01 16:18:00   Saturday  TENDERLOIN   
679267  679267 2005-09-30 23:01:00     Friday    NORTHERN   
683721  683721 2005-09-12 17:30:00     Monday  TENDERLOIN   
683795  683795 2005-09-12 13:53:00     Monday  TENDERLOIN   
683968  683968 2005-09-04 23:30:00     Sunday    NORTHERN   
683970  683970 2005-09-04 23:30:00     Sunday    NORTHERN   
686013  686013 2005-08-30 20:15:00    Tuesday  TENDERLOIN   
686014  686014 2005-08-30 20:15:00    Tuesday  TENDERLOIN   
686015  686015 2005-08-30 20:15:00    Tuesday  TENDERLOIN   
687015  687015 2005-08-21 05:30:00     Sunday    NORTHERN   
687357  687357 2005-08-20 09:30:00   Saturday  TENDERLOIN   
688886  688886 2005-08-16 09:25:00    Tuesday  TENDERLOIN   
689209  689209 2005-08-15 12:00:00     Monday  TENDERLOIN   
693007  693007 2005-07-22 22:15:00     Friday     TARAVAL   
705820  705820 2005-05-23 02:00:00     Monday     CENTRAL   
707005  707005 2005-05-13 00:18:00     Friday     CENTRAL   
709457  709457 2005-04-29 18:55:00     Friday  TENDERLOIN   
...        ...                 ...        ...         ...   
790277  790277 2004-03-27 12:48:00   Saturday    NORTHERN   
790278  790278 2004-03-27 12:48:00   Saturday    NORTHERN   
796525  796525 2004-02-27 16:39:00     Friday    NORTHERN   
796526  796526 2004-02-27 16:39:00     Friday    NORTHERN   
798831  798831 2004-02-15 02:00:00     Sunday    SOUTHERN   
798832  798832 2004-02-15 02:00:00     Sunday    SOUTHERN   
798941  798941 2004-02-14 21:10:00   Saturday    RICHMOND   
798942  798942 2004-02-14 21:10:00   Saturday    RICHMOND   
798943  798943 2004-02-14 21:10:00   Saturday    RICHMOND   
800668  800668 2004-02-10 20:30:00    Tuesday    RICHMOND   
800670  800670 2004-02-10 20:30:00    Tuesday    RICHMOND   
807432  807432 2004-01-12 05:30:00     Monday    SOUTHERN   
807434  807434 2004-01-12 05:30:00     Monday    SOUTHERN   
814294  814294 2003-12-03 16:30:00  Wednesday    SOUTHERN   
822978  822978 2003-10-22 01:38:00  Wednesday    NORTHERN   
822979  822979 2003-10-22 01:38:00  Wednesday    NORTHERN   
822980  822980 2003-10-22 01:38:00  Wednesday    NORTHERN   
822981  822981 2003-10-22 01:38:00  Wednesday    NORTHERN   
824015  824015 2003-10-12 17:46:00     Sunday    NORTHERN   
824016  824016 2003-10-12 17:46:00     Sunday    NORTHERN   
824017  824017 2003-10-12 17:46:00     Sunday    NORTHERN   
825266  825266 2003-10-09 14:51:00   Thursday     MISSION   
825915  825915 2003-10-07 23:45:00    Tuesday     BAYVIEW   
827026  827026 2003-09-28 10:00:00     Sunday    SOUTHERN   
829626  829626 2003-09-22 13:50:00     Monday   INGLESIDE   
830647  830647 2003-09-12 23:00:00     Friday    NORTHERN   
830648  830648 2003-09-12 23:00:00     Friday    NORTHERN   
836254  836254 2003-08-16 17:30:00   Saturday    SOUTHERN   
836255  836255 2003-08-16 17:30:00   Saturday    SOUTHERN   
861864  861864 2003-04-12 11:30:00   Saturday     TARAVAL   

                                          Address      X   Y  
664733              7THSTNORTH ST / MCALLISTER ST -120.5  90  
664734              7THSTNORTH ST / MCALLISTER ST -120.5  90  
664826             GEARY BL / AVENUE OF 

In [36]:
test_set[test_set.X < -121].X.describe()

count    884186.000000
mean       -122.422858
std           0.025345
min        -122.513642
25%        -122.433071
50%        -122.416517
75%        -122.406959
max        -122.364751
Name: X, dtype: float64

### Conclusion of data exploring
Everything looks fine except there is a special coordinate -120.5 and 90. This looks like a default coordinate when GPS is not working. try to recover the coordinate

In [75]:
missing_coordinate_addresses = (train_set[train_set.Y == 90].Address.append(test_set[test_set.Y == 90].Address)).unique()

In [77]:
len(missing_coordinate_addresses)

59

In [86]:
train_set[train_set.Address.isin(missing_coordinate_addresses) & (train_set.Y != 90)]

Dates        Category  \
715474 2005-03-24 22:40:00         ASSAULT   
722391 2005-02-21 02:30:00  OTHER OFFENSES   
741420 2004-10-30 07:39:00  OTHER OFFENSES   
745970 2004-10-11 19:35:00  OTHER OFFENSES   
754788 2004-08-22 02:35:00  OTHER OFFENSES   
772427 2004-05-28 18:00:00   VEHICLE THEFT   
811431 2003-11-16 21:41:00  OTHER OFFENSES   

                                     Descript DayOfWeek PdDistrict  \
715474          ATTEMPTED HOMICIDE WITH A GUN  Thursday  INGLESIDE   
722391                      TRAFFIC VIOLATION    Monday    TARAVAL   
741420  DRIVERS LICENSE, SUSPENDED OR REVOKED  Saturday    BAYVIEW   
745970                      TRAFFIC VIOLATION    Monday    BAYVIEW   
754788  DRIVERS LICENSE, SUSPENDED OR REVOKED    Sunday    MISSION   
772427                           STOLEN TRUCK    Friday   SOUTHERN   
811431  DRIVERS LICENSE, SUSPENDED OR REVOKED    Sunday    BAYVIEW   

           Resolution                                  Address           X  \
715474           NONE              INTERSTATE280 HY / OCEAN AV -122.449177   
722391  ARREST, CITED             INTERSTATE280 HY / GENEVA AV -122.447749   
741420  ARREST, CITED                  I-280 / CESAR CHAVEZ ST -122.391983   
745970  ARREST, CITED        JAMES LICK FREEWAY HY / SILVER AV -122.405088   
754788  ARREST, CITED  JAMES LICK FREEWAY HY / CESAR CHAVEZ ST -122.403227   
772427           NONE                     BRYANT ST / SPEAR ST -122.388063   
811431  ARREST, CITED                  I-280 / PENNSYLVANIA AV -122.392561   

                Y  
715474  37.722963  
722391  37.721155  
741420  37.750082  
745970  37.731239  
754788  37.749464  
772427  37.787204  
811431  37.751754

In [87]:
test_set[test_set.Address.isin(missing_coordinate_addresses) & (test_set.Y != 90)]

Id               Dates  DayOfWeek PdDistrict  \
723014  723014 2005-02-19 14:00:00   Saturday  INGLESIDE   
801637  801637 2004-02-01 14:29:00     Sunday    MISSION   
810402  810402 2003-12-20 21:05:00   Saturday  INGLESIDE   
834789  834789 2003-08-27 15:00:00  Wednesday   SOUTHERN   

                                        Address           X          Y  
723014              INTERSTATE280 HY / OCEAN AV -122.449177  37.722963  
801637  JAMES LICK FREEWAY HY / CESAR CHAVEZ ST -122.405097  37.749110  
810402     JAMES LICK FREEWAY HY / BAY SHORE BL -122.399602  37.720922  
834789      SPEAR ST / THE EMBARCADERO SOUTH ST -122.388196  37.788005

#### Able to recover some coordinates from test/train data

# Using svm

In [29]:
from sklearn import svm
from sklearn import preprocessing

## Casting and feature scaling